File: python/example/sort/service/heapsort_service.py
from otterop.io.console import Console
from otterop.lang.array import Array
from otterop.lang.string import String

class HeapsortService:

    def __init__(self):
        pass
    def swap(self, array, from_idx, to_idx):
        tmp = array.get(to_idx)
        array.set(to_idx, array.get(from_idx))
        array.set(from_idx, tmp)

    def heapify(self, array, parent, size):
        largest = parent
        left_child = parent * 2 + 1
        right_child = parent * 2 + 2
        if left_child < size:
            left_s = array.get(left_child)
            if left_s.compare_to(array.get(largest)) > 0:
                largest = left_child
        if right_child < size:
            right_s = array.get(right_child)
            if right_s.compare_to(array.get(largest)) > 0:
                largest = right_child
        if largest != parent:
            self.swap(array, parent, largest)
            self.heapify(array, largest, size)

    def sort(self, array, from_idx, to_idx):
        n = array.size()
        i = n // 2 - 1
        while i >= 0:
            self.heapify(array, i, n)
            i -= 1
        i = n - 1
        while i >= 0:
            self.swap(array, 0, i)
            self.heapify(array, 0, i)
            i -= 1

File: ts/example/sort/service/HeapsortService.ts
import { Array } from '@otterop/lang/array';
import { String } from '@otterop/lang/string';

export class HeapsortService {

    public constructor() {
    }

    swap(array : Array<String>, fromIdx : number, toIdx : number) : void {
        let tmp : String = array.get(toIdx);
        array.set(toIdx, array.get(fromIdx));
        array.set(fromIdx, tmp);
    }

    heapify(array : Array<String>, parent : number, size : number) : void {
        let largest : number = parent;
        let leftChild : number = parent * 2 + 1;
        let rightChild : number = parent * 2 + 2;
        
        if (leftChild < size) {
            let leftS : String = array.get(leftChild);
            
            if (leftS.compareTo(array.get(largest)) > 0) {
                largest = leftChild;
            }
        }
        
        if (rightChild < size) {
            let rightS : String = array.get(rightChild);
            
            if (rightS.compareTo(array.get(largest)) > 0) {
                largest = rightChild;
            }
        }
        
        if (largest != parent) {
            this.swap(array, parent, largest);
            this.heapify(array, largest, size);
        }
    }

    public sort(array : Array<String>, fromIdx : number, toIdx : number) : void {
        let n : number = array.size();
        for (let i : number = Math.floor(n / 2) - 1; i >= 0; i--) {
            this.heapify(array, i, n);
        }
        for (let i : number = n - 1; i >= 0; i--) {
            this.swap(array, 0, i);
            this.heapify(array, 0, i);
        }
    }
}

